test/Microsoft.VisualStudio.SDK.Analyzers.Tests/VSSDK008ThreadAffinitizedMEFConstructionTests.cs

```diff
@@ -0,0 +1,638 @@
+using System.ComponentModel.Composition;
+
+[Export]
+internal class C
+{
+    bool? MyProperty = null;
+    public C()
+    {
+        MyProperty = false;
+    }
+}
+";
+
+        await Verify.VerifyAnalyzerAsync(test);
+    }
+
+    [Fact]
+    public async Task ObjectOrCollectionInitializer_Benign_NoWarning()
+    {
+        var test = /* lang=c#-test */ @"
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Utilities;
+
+[Export]
+[Name(PartName)]
+[Order(Before = ""default"")]
+internal class C
+{
+    public const string PartName = ""PartName"";
+    public C()
+    {
+    }
+}
+";
+
+        await Verify.VerifyAnalyzerAsync(test);
+    }
+
+    [Fact]
+    public async Task Constructor_MainThreadRequired_LocalFunction_Flagged()
+    {
+        var test = /* lang=c#-test */ @"
+using System.ComponentModel.Composition;
+
+[Export]
+class C
+{
+    public C()
+    {
+        LocalFunction();
+
+        void LocalFunction()
+        {
+            [|Microsoft.VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread()|];
+        }
+    }
+}";
+
+        await Verify.VerifyAnalyzerAsync(test);
+    }
+
+    /// <summary>
+    /// This test case showcases a false positive, where a file-and-forget function is flagged.
+    /// It's OK to access UI thread within the local function which executes asynchronously and
+    /// MEF part construction does not join on this async work.
+    /// </summary>
+    [Fact(Skip = "False positive.")]
+    public async Task Constructor_MainThreadRequired_LocalFunction_FireAndForget_Flagged_FalsePositive()
+    {
+        var test = /* lang=c#-test */ @"
+using Microsoft.VisualStudio.Threading;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+[Export]
+class C
+{
+    [ImportingConstructor]
+    public C(JoinableTaskContext joinableTaskContext)
+    {
+        LocalFunctionAsync().Forget();
+
+        async Task LocalFunctionAsync()
+        {
+            await joinableTaskContext.Factory.SwitchToMainThreadAsync(); // False positive reported here
+            return;
+        }
+    }
+}";
+
+        await Verify.VerifyAnalyzerAsync(test);
+    }
+
+    /// <summary>
+    /// This test showcases a false negative, where VSSDK013 does not report
+    /// UI thread dependency inside the async lambda. To reduce number of false positives,
+    /// we assume that lambdas are executed asynchronously.
+    /// In this example, execution joins the lambda, effectively blocking UI thread.
+    /// </summary>
+    [Fact(Skip = "False negative")]
+    public async Task JoiningAsyncLambda_MainThreadAsserted_NoWarning_FalseNegative()
+    {
+        var test = /* lang=c#-test */ @"
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Threading;
+
+[Export]
+class C
+{
+    private readonly AsyncLazy<object> _o;
+
+    [ImportingConstructor]
+    public C(JoinableTaskContext joinableTaskContext)
+    {
+        var s = joinableTaskContext.Factory.Run(async () =>
```

Comment:
Calling `JTF.Run` itself is a Bad Thing in a MEF part, since MEF parts should activate quickly and synchronously, and such a call implies that it is either slow or has thread affinity.

Reply:
Good point, I changed the example